### PR TITLE
fix: New external governor module proposal for emergency DVM actions

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

This PR adds a new emergency exit path to the dvm to recover from the voting system breaking.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/oracle/implementation/GovernorV2.sol [hunk 1]
```diff
@@ -179,6 +168,39 @@ contract GovernorV2 is MultiRole, Lockable {
         return proposals[id];
     }
 
+    /****************************************
+     *       EMERGENCY ADMIN FUNCTIONS      *
+     ****************************************/
+
+    function proposeEmergencyAction(Transaction[] memory transactions)
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/GovernorV2.sol:175]: function called by the `SafteyModule`.

File: packages/core/contracts/oracle/implementation/GovernorV2.sol [hunk 2]
```diff
@@ -208,7 +230,42 @@ contract GovernorV2 is MultiRole, Lockable {
...
 
+    function _getSafetyModule() private view returns (SafetyModuleInterface) {
+        return SafetyModuleInterface(getMember(uint256(Roles.EmergencyProposer)));
+    }
+
     function _getIdentifierWhitelist() private view returns (IdentifierWhitelistInterface) {
         return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
     }
+
+    function _appendTransactionsToProposal(Transaction[] memory transactions, Proposal storage proposal) internal {
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/GovernorV2.sol:241]: logic that was internal within the two other functions now pulled to be re-used.

File: packages/core/contracts/oracle/implementation/SafetyModule.sol [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import "./GovernorV2.sol"; //todo: swap this to using an interface.
+import "./Constants.sol";
+import "../interfaces/StakerInterface.sol";
+import "./AdminIdentifierLib.sol";
+import "../../common/implementation/Lockable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract SafteyModule is Ownable, Lockable {
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/SafetyModule.sol:14]: now contract that lets someone propose an emergency action and ratify it. Note that this requires that the `emergencyActionAuthenticator` executes the action, thereby removing a lot of nasty things that could be done if this was a separate governance action, such as front running the deployment or the related attacks.

**Comment by chrismaree** [comment]
Main differences to what I did last time:
1. it integrates with the `Governor`, rather than the `Voting` contract
2. it is more easily upgradable (we can easily set the `SafetyModule` to a new implementation via governance thereby changing emergency logic via an upgrade of this one bit of logic).
3. it contains the notion of `emergencyActionAuthenticator` that must ratify an emergency action. This will be set to a multisig and is the only account that can execute a governance action. you still require a sufficient quorum to be hit before something can be executed but only this account can actually execute the emergency action. this gives us some control that this upgrade path cant break the dvm, while still keeping it decentralized and governed by the token holders.

File: packages/core/contracts/oracle/implementation/SafetyModule.sol [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    function signalOnEmergencyProposal(uint256 id) public {
+        EmergencyProposal storage emergencyProposal = emergencyProposals[id];
+
+        require(emergencyProposal.accountSignaled[msg.sender] == 0);
+
+        emergencyProposal.accountSignaled[msg.sender] = getVoterStake(msg.sender);
+        emergencyProposal.cumulativeSignaled += emergencyProposal.accountSignaled[msg.sender];
+    }
```
Comment (mrice32) [inline packages/core/contracts/oracle/implementation/SafetyModule.sol:74]: If users are able to unstake, they could replicate their votes, right? I could stake with account1, signal, withdraw, restake with a second account, signal, etc.

File: packages/core/contracts/oracle/implementation/SafetyModule.sol [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        address proposer;
+        uint64 time;
+        uint256 lockedBond;
+        uint256 cumulativeSignaled;
+        mapping(address => uint256) accountSignaled;
+        bool ratified;
+    }
+    EmergencyProposal[] public emergencyProposals;
+
+    uint256 emergencyActionThreshold = 0.5e18;
```
Comment (mrice32) [inline packages/core/contracts/oracle/implementation/SafetyModule.sol:31]: Similar to the last implementation, a percentage cutoff means that during migrations, the system is more vulnerable and the multisig would have substantial power (i.e. collusion between multisig and another address could take over the system), right?

**Review by mrice32 — COMMENTED** [review]
Just a few initial comments. Looks great!

File: packages/core/contracts/oracle/implementation/SafetyModule.sol (hunk 5 above)
Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/SafetyModule.sol:31]: correct.

File: packages/core/contracts/oracle/implementation/SafetyModule.sol (hunk 4 above)
Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/SafetyModule.sol:74]: ye, you right. in the implementation where this was within the voting, contract we'd block a voter from unstaking if they'd signed on an emergency action. options would be to block the staker from unstaking within the voting contract if they'd signed on an emergency action?

Comment (chrismaree) [reply packages/core/contracts/oracle/implementation/SafetyModule.sol]: I've updated the proposal with this minimal change. basically, we now track a counter of how many proposals an account is signaled on. you can only Unstake if this is set to 0, thereby forcing you to cancel your signal on everything before you can unstake. This does add a small gas cost to executing an unstake but it should not be too much as there are only two small read intra-contract calls (call to the finder and a call to the safety module).

File: packages/core/contracts/oracle/implementation/Staker.sol [hunk 6]
```diff
@@ -302,6 +309,12 @@ abstract contract Staker is StakerInterface, Ownable, Lockable {
         return block.timestamp;
     }
 
+    function isVoterActivelySignaledOnEmergencyAction(address voterAddress) public view returns (bool) {
```
Comment (chrismaree) [inline packages/core/contracts/oracle/implementation/Staker.sol:312]: see here for the intra-calls to block if you are currently signed on a safety action.

**Comment by chrismaree** [comment]
closed in favor of: https://github.com/UMAprotocol/protocol/pull/4128

**Pull request closed without merging**